### ISSUE TITLE
overlay networking with userland-proxy disabled prevents port exposure

### ISSUE THREAD

**Issue by cgeers** [docker/docker#22741]
I've run across an issue with the combination of docker-engine's builtin overlay networking when userland-proxy is disabled.  The root of this issue comes about from the fact that when a container is attached to an overlay network, a new bridge network is also created dynamically to provide a path for outside connectivity.  This bridge is currently named `docker_gwbridge`.  This bridge, because it is used by *all* containers attach to *any* overlay network present on the host, rightfully so, is created with an *implicit* ICC=false configuration, even when ICC is enabled in docker-engine. (See #19298 for details). 

This seems like a reasonable configuration, however, the network isolation provided by the lack of ICC on docker_gwbridge is overly aggressive. So much so, that it breaks other longstanding docker functionality such as exposing a port provided by the container on the host when userland-proxy is disabled.  

For background, when userland-proxy is disabled, docker-engine will inject DNAT and MASQUERADE rules into the iptables `nat` table to allow the kernel to facilitate port forwarding from host to container, in essence, _simulating_ port exposure on the host by a container using the `-p <port>:<port>` option. For obvious performance reasons, this is *much* preferable to using the the userland tcp proxy process. 

If we look at how this implicit ICC=false configuration of docker_gwbridge works in iptables when in userland-proxy is disabled we see we have the following DOCKER-ISOLATION chain created the `filter` table:

```
Chain DOCKER-ISOLATION (1 references)
num   pkts bytes target     prot opt in     out     source               destination         
1        0     0 DROP       all  --  docker0 docker_gwbridge  0.0.0.0/0            0.0.0.0/0           
2        0     0 DROP       all  --  docker_gwbridge docker0  0.0.0.0/0            0.0.0.0/0           
3        0     0 RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0 
```  

This is intended to prevent the containers on the docker_gwbridge from having free reign of containers the docker0 bridge, and vice versa.  This makes total sense, and definitely is the intended behavior of ICC=false on docker_gwbridge.  If we look at the default configuration of  theFORWARD chain in the `filter` table, we'll see the issue:

```
Chain FORWARD (policy ACCEPT 0 packets, 0 bytes)
num   pkts bytes target     prot opt in     out     source               destination         
1        0     0 DOCKER-ISOLATION  all  --  *      *       0.0.0.0/0            0.0.0.0/0           
2        0     0 DOCKER     all  --  *      docker_gwbridge  0.0.0.0/0            0.0.0.0/0           
3        0     0 ACCEPT     all  --  *      docker_gwbridge  0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED
4        0     0 ACCEPT     all  --  docker_gwbridge !docker_gwbridge  0.0.0.0/0            0.0.0.0/0           
5        0     0 DOCKER     all  --  *      docker0  0.0.0.0/0            0.0.0.0/0           
6        0     0 ACCEPT     all  --  *      docker0  0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED
7        0     0 ACCEPT     all  --  docker0 !docker0  0.0.0.0/0            0.0.0.0/0           
8        0     0 ACCEPT     all  --  docker0 docker0  0.0.0.0/0            0.0.0.0/0           
9        0     0 DROP       all  --  docker_gwbridge docker_gwbridge  0.0.0.0/0            0.0.0.0/0  
```

Here we see that the DOCKER-ISOLATION chain is referenced first, *before* we have a chance to get DOCKER chain, whose (sole?) purpose is *alllow* DNAT traffic ingress on docker_gwbridge (created by the port exposure). The placement of the reference to the DOCKER-ISOLATION chain in the FORWARD chain seems to be incorrect -- it should *not* be first in the list. 

My proposed change is to move the reference to the DOCKER-ISOLATION chain down to rule number 4 in the FORWARD chain, allowing the 2,3,4 rules to become 1,2,3 respectively. Like so:

```
Chain FORWARD (policy ACCEPT 0 packets, 0 bytes)
num   pkts bytes target     prot opt in     out     source               destination         
1        0     0 DOCKER     all  --  *      docker_gwbridge  0.0.0.0/0            0.0.0.0/0           
2        0     0 ACCEPT     all  --  *      docker_gwbridge  0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED
3        0     0 ACCEPT     all  --  docker_gwbridge !docker_gwbridge  0.0.0.0/0            0.0.0.0/0           
4        0     0 DOCKER-ISOLATION  all  --  *      *       0.0.0.0/0            0.0.0.0/0           
5        0     0 DOCKER     all  --  *      docker0  0.0.0.0/0            0.0.0.0/0           
6        0     0 ACCEPT     all  --  *      docker0  0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED
7        0     0 ACCEPT     all  --  docker0 !docker0  0.0.0.0/0            0.0.0.0/0           
8        0     0 ACCEPT     all  --  docker0 docker0  0.0.0.0/0            0.0.0.0/0           
9        0     0 DROP       all  --  docker_gwbridge docker_gwbridge  0.0.0.0/0            0.0.0.0/0   
```

Making this change has the following effect:

1) Just as before the change, *all* containers on the docker_gwbridge (regardless of the overlay network they are on) can *not* ping or connect to layer4 ports of other containers on the docker_gwbridge bridge, *including* ports exposed and mapped to the host by containers on this bridge. This behavior is exactly as it was before the change, and is still reasonable since the overlay network itself presents a path for containers to communicate, assuming the engineer places the containers together in the same overlay network in the first place.

2) Just as before the change, containers on the docker0 bridge (not part of any overlay) *can not* connect arbitrarily to any layer4 port that happens to be listening and exposed (but not mapped to the host) on containers attached to any overlay network. As before, ICMP ping is also blocked over this path. This again is reasonable and correct behavior given the implicit ICC=false configuration of docker_gwbridge when it's created.

3) Containers on the docker0 bridge (not part of any overlay) *can* connect to other containers on an overlay network that are configured to expose ports *and* those ports are mapped to a host port.

TL;DR:
Even when ICC is false (in this example, it globally true, but implicitly false for docker_gwbridge) the user should still be able to configure containers on overlay networks to expose ports and map them to the host to allow other containers to connect.

Here's how to reproduce: 

**Output of `docker version`:**

```
Docker version 1.11.1, build 5604cbe
```

**Output of `docker info`:**

```
Containers: 9
 Running: 8
 Paused: 0
 Stopped: 1
Images: 9
Server Version: 1.11.1
Storage Driver: overlay
 Backing Filesystem: extfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: overlay bridge null host
Kernel Version: 4.4.10-1.el7.elrepo.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 992.1 MiB
Name: localhost.local
ID: BQMV:ALZI:2JE3:EZLW:BTX2:4TUM:5U4Z:OJGY:NXAA:DTHK:BLUU:PRQP
Docker Root Dir: /var/lib/docker
Debug mode (client): false
Debug mode (server): true
 File Descriptors: 71
 Goroutines: 116
 System Time: 2016-05-14T12:21:17.052722898-04:00
 EventsListeners: 0
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-ip6tables is disabled
Labels:
Cluster store: consul://localhost:8080
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

vagrant + virtualbox

**Steps to reproduce the issue:**

1. Configure docker engine for overlay networking, have ICC enabled, and userland-proxy disabled
2. Create a container (container1) on an overlay network that exposes a listening TCP port mapped to the host
3. Create a new container (container2) that attached to the docker0 bridge

**Describe the results you received:**

container2 is *unable* to telnet to the TCP port exposed and mapped to the host by container1 

**Describe the results you expected:**
 
container2 is *able* to telnet to the TCP port exposed and mapped to the host by container1

**Additional information you deem important (e.g. issue happens only occasionally):**

This is reproducible consistently. This issue is caused by an overly aggressive ordering iptables rules. Specifically the placement of the DOCKER-ISOLATION rule at the top of the FORWARD chain prevents this from working.

**Comment by thaJeztah** [comment]
Thanks for a very detailed report :+1:

**Comment by thaJeztah** [comment]
ping @mavenugo could you have a look at this one?

**Comment by mavenugo** [comment]
@cgeers thanks for the good analysis and detailed understanding on the `gw_bridge` usage. Yes, `userland-proxy = false` handling is tricky & has some `iptables` conflicts. Can you please confirm if this is an issue only for container1 attached to docker0 in the same host as container2 attached to the overlay network ?
With the multiple network support, one can `docker network connect` contaienr1 and container2 to a common bridge network and they can access each other without the need for port-mapping.

**Comment by cgeers** [comment]
Hi @mavenugo 

>  Can you please confirm if this is an issue only for container1 attached to docker0 in the same host as container2 attached to the overlay network ?

yes, that the primary use case. Ports mapped to the host by containers in an overlay are effectively configured to publish these ports for others to consume. Several iptables rules (DNAT and MASQUERADE in the nat table) are injected to support such a container's port mapping.  However, these rules are seemingly *never* reached due to the aggressive placement of the reference to the DOCKER-ISOLATION chain within the FORWARD chain.

> With the multiple network support, one can docker network connect contaienr1 and container2 to a common bridge network and they can access each other without the need for port-mapping.

that's certainly true, and I've toyed with doing this to some degree.  Firstly, it somewhat defeats the purpose of isolating container traffic into different networks the first place.  Secondly, doing this can lead to some interesting routing scenarios. For example, my two container demonstration above, you would attach container1 (on the overlay and `gw_bridge`) to the docker0 bridge.  Doing this creates a third interface on container1, two of which are veths attached to host bridges, and only one of those two interfaces is the path for the default route.  At this point it can be easy to get yourself into an asymmetric routing scenario, which the host's netfilter doesn't take kindly to. Suffice it to say, my experience with going this path hasn't been great. This, compounded with the fact that I'm seemingly weakening the network security between containers all to work around what appears (at least to me) to be a mis-ordering of rules in the FORWARD chain is why I haven't pursued this path much. 

Thanks for your thoughts.

**Comment by mavenugo** [comment]
@cgeers From your analysis, it would have been clear that even though `gw_bridge` is a special network to provide external connectivity, it is NOT anyway special to the bridge driver that owns the `DOCKER-ISOLATION` rules. Hence the placement of rules as you suggested will not work. Also, ignoring the overlay use-case, This will be the exact same behavior for all other networks backed by bridge driver as well. Hence, any solution to this must be generic that solves the common problem.

Also, this is a problem specific to `userland-proxy=false` and we have other bigger issues (https://github.com/docker/docker/issues/5618) to be solved first before us recommending our users to use that. Though it is ortogonal, I just want to bring this to your attention and a possible reason on why your particular issue is not encountered by others (since most of us keep the proxy running).

**Comment by cgeers** [comment]
Hi @mavenugo Yes, I believe I understand what you mean. In my example, my use of the overlay and thus `gw_bridge` is not strictly required.  In fact, neither is ICC (implicit or otherwise), correct? I haven't yet verified it, but it seems that this behavior I describe should be reproducible with any two bridge networks when userland-proxy=false. Does this seem plausible?

**Comment by mavenugo** [comment]
@cgeers thats correct. this issue is not specific to overlay network. It is to do with `userland-proxy=false` and yes, we should be able to reproduce this with 2 bridge networks with `userland-proxy=false`.
cc @aboch

**Comment by cirocosta** [comment]
Hey, I suppose many things changed from May 2016; is this still a thing? Or in the area of not using userland-proxy nothing changed?

Thx!

**Comment by jowi24** [comment]
I still experience this problem with docker 18.06.1-ce

**Comment by akerouanton** [comment]
I can't reproduce this issue with current HEAD, v27.0.0 and v23.0.0. Despite the last comment, I believe this was fixed circa 2016-2018 (probably by https://github.com/moby/libnetwork/pull/1398), so let me close this issue.
